shell/renderer/api/electron_api_context_bridge.cc

```diff
@@ -687,33 +687,80 @@ void ExposeAPIInWorld(v8::Isolate* isolate,
     return;
   }
 
-  v8::Local<v8::Context> electron_isolated_context =
-      frame->GetScriptContextFromWorldId(args->isolate(),
-                                         WorldIDs::ISOLATED_WORLD_ID);
+  context_bridge::ObjectCache object_cache;
 
-  {
-    context_bridge::ObjectCache object_cache;
-    v8::Context::Scope target_context_scope(target_context);
+  v8::MaybeLocal<v8::Value> maybe_proxy = PassValueToOtherContext(
+      source_context, target_context, api, source_context->Global(),
+      &object_cache, false, 0, BridgeErrorTarget::kSource);
+  if (maybe_proxy.IsEmpty())
+    return;
+  auto proxy = maybe_proxy.ToLocalChecked();
 
-    v8::MaybeLocal<v8::Value> maybe_proxy = PassValueToOtherContext(
-        electron_isolated_context, target_context, api,
-        electron_isolated_context->Global(), &object_cache, false, 0,
-        BridgeErrorTarget::kSource);
-    if (maybe_proxy.IsEmpty())
-      return;
-    auto proxy = maybe_proxy.ToLocalChecked();
+  if (base::FeatureList::IsEnabled(features::kContextBridgeMutability)) {
+    global.Set(key, proxy);
+    return;
+  }
 
-    if (base::FeatureList::IsEnabled(features::kContextBridgeMutability)) {
-      global.Set(key, proxy);
-      return;
-    }
+  if (proxy->IsObject() && !proxy->IsTypedArray() &&
+      !DeepFreeze(proxy.As<v8::Object>(), target_context))
+    return;
 
-    if (proxy->IsObject() && !proxy->IsTypedArray() &&
-        !DeepFreeze(proxy.As<v8::Object>(), target_context))
-      return;
+  global.SetReadOnlyNonConfigurable(key, proxy);
+}
 
-    global.SetReadOnlyNonConfigurable(key, proxy);
+// Attempt to get the target context based on the current context.
+//
+// For render frames, this is either the main world (0) or an arbitrary
+// world ID. For service workers, Electron only supports one isolated
+// context and the main worker context. Anything else is invalid.
+v8::MaybeLocal<v8::Context> GetTargetContext(v8::Isolate* isolate,
+                                             const int world_id) {
+  v8::Local<v8::Context> source_context = isolate->GetCurrentContext();
+  v8::MaybeLocal<v8::Context> maybe_target_context;
+
+  blink::ExecutionContext* execution_context =
+      blink::ExecutionContext::From(source_context);
```

Comment:
Do we know how expensive this is, it might be worth passing this down through arg binds instead of recomputing in the function wrapper each time. i.e. calc once and then all transitive "bridged" things retain that state. A context can't transition from "window" to "worker" so we should be good to cache like that

Reply:
Good call out, I was not thinking about the recursion when initially adding this.

I refactored `PassValueToOtherContext` to call an inner `PassValueToOtherContextInner` ([b06dfb7](https://github.com/electron/electron/pull/44411/commits/b06dfb722e9f060f37a180d1b71f0b4b97b19dd8)). This allows us to setup any context needed ahead of time, like the execution context. It has the added bonus of enforcing a new object cache for each call which addresses https://github.com/electron/electron/pull/44411#discussion_r1861378754.